zkevm-circuits/src/evm_circuit/execution/sha3.rs

```diff
@@ -0,0 +1,201 @@
+            memory_address.address(),
+            0.expr(), // dst_addr for CopyDataType::RlcAcc is 0.
+            memory_address.length(),
+            rlc_acc.expr(),
+            cb.curr.state.rw_counter.expr() + cb.rw_counter_offset(),
+            copy_rwc_inc.expr(),
+        );
+        cb.keccak_table_lookup(rlc_acc.expr(), memory_address.length(), sha3_rlc.expr());
+
+        let memory_expansion = MemoryExpansionGadget::construct(
+            cb,
+            cb.curr.state.memory_word_size.expr(),
+            [memory_address.address()],
+        );
+        let memory_copier_gas = MemoryCopierGasGadget::construct(
+            cb,
+            memory_address.length(),
+            memory_expansion.gas_cost(),
+        );
+
+        let step_state_transition = StepStateTransition {
+            rw_counter: Transition::Delta(cb.rw_counter_offset() + copy_rwc_inc.expr()),
+            program_counter: Transition::Delta(1.expr()),
+            stack_pointer: Transition::Delta(1.expr()),
+            memory_word_size: Transition::To(memory_expansion.next_memory_word_size()),
+            gas_left: Transition::Delta(
+                -(OpcodeId::SHA3.constant_gas_cost().expr() + memory_copier_gas.gas_cost()),
+            ),
+            ..Default::default()
+        };
+        let same_context = SameContextGadget::construct(cb, opcode, step_state_transition);
+
+        Self {
+            same_context,
+            memory_address,
+            sha3_rlc,
+            copy_rwc_inc,
+            rlc_acc,
+            memory_expansion,
+            memory_copier_gas,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        tx: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let [memory_offset, size, sha3_output] =
+            [step.rw_indices[0], step.rw_indices[1], step.rw_indices[2]]
+                .map(|idx| block.rws[idx].stack_value());
+        let memory_address =
+            self.memory_address
+                .assign(region, offset, memory_offset, size, block.randomness)?;
+        self.sha3_rlc
+            .assign(region, offset, Some(sha3_output.to_le_bytes()))?;
+
+        let key = (tx.id, call.id, step.gas_left);
+        let copy_rwc_inc = block
+            .copy_events
+            .get(&key)
+            .unwrap()
+            .steps
+            .first()
+            .map_or(F::zero(), |cs| F::from(cs.rwc_inc_left));
+        self.copy_rwc_inc
+            .assign(region, offset, Some(copy_rwc_inc))?;
+
+        let values = block
+            .copy_events
+            .get(&key)
+            .unwrap()
+            .steps
+            .iter()
+            .filter(|s| s.rw.is_write())
+            .map(|s| s.value)
+            .collect::<Vec<u8>>();
```

Comment:
```suggestion
use eth_types::ToScalar // move this to the top of the file.
self.copy_rwc_inc.assign(region, offset, size.to_scalar())?;

 let values: Vec<u8> = (3..3 + (size.low_u64() as usize))
      .map(|i| block.rws[step.rw_indices[i]].memory_value())
      .collect();
```

Reply:
>(1) i mentioned "let key = (tx.id, call.id, step.program_counter as usize);" may be not uniq enough.
We can get rid of the key altogether here (and probably in all the other copy table lookups as well), but just directly calculating `copy_rwc_inc` and `values`.